dbms/src/Storages/DeltaMerge/LateMaterializationBlockInputStream.cpp

```diff
@@ -0,0 +1,134 @@
+#include <Storages/DeltaMerge/ReadUtil.h>
+
+#include <algorithm>
+
+namespace DB::DM
+{
+
+LateMaterializationBlockInputStream::LateMaterializationBlockInputStream(
+    const ColumnDefines & columns_to_read,
+    BlockInputStreamPtr filter_column_stream_,
+    SkippableBlockInputStreamPtr rest_column_stream_,
+    const BitmapFilterPtr & bitmap_filter_,
+    const String & req_id_)
+    : header(toEmptyBlock(columns_to_read))
+    , filter_column_stream(filter_column_stream_)
+    , rest_column_stream(rest_column_stream_)
+    , bitmap_filter(bitmap_filter_)
+    , log(Logger::get(NAME, req_id_))
+{}
+
+Block LateMaterializationBlockInputStream::readImpl()
+{
+    Block filter_column_block;
+    FilterPtr filter = nullptr;
+
+    // Until non-empty block after filtering or end of stream.
+    while (true)
+    {
+        filter_column_block = filter_column_stream->read(filter, true);
+
+        // If filter_column_block is empty, it means that the stream has ended.
+        // No need to read the rest_column_stream, just return an empty block.
+        if (!filter_column_block)
+            return filter_column_block;
+
+        RUNTIME_CHECK_MSG(filter, "Late materialization meets unexpected null filter");
+
+        // Get mvcc-filter
+        size_t rows = filter_column_block.rows();
+        mvcc_filter.resize(rows);
+        bool all_match = bitmap_filter->get(mvcc_filter, filter_column_block.startOffset(), rows);
+        if (!all_match)
+        {
+            // if mvcc-filter is all match, use filter directly
+            // else use `mvcc-filter & filter` to get the final filter
+            std::transform(mvcc_filter.cbegin(), mvcc_filter.cend(), filter->cbegin(), filter->begin(), [](const UInt8 a, const UInt8 b) { return a != 0 && b != 0; });
+        }
+
+        if (size_t passed_count = countBytesInFilter(*filter); passed_count == 0)
+        {
+            // if all rows are filtered, skip the next block of rest_column_stream
+            if (rest_column_stream->skipNextBlock() == 0)
+            {
+                // if we fail to skip, we need to call read() of rest_column_stream, but ignore the result
+                // NOTE: skipNextBlock() return 0 only if failed to skip or meets the end of stream,
+                //       but the filter_column_stream doesn't meet the end of stream
+                //       so it is an unexpected behavior.
+                rest_column_stream->read();
+                LOG_ERROR(log, "Late materialization skip block failed, at start_offset: {}, rows: {}", filter_column_block.startOffset(), filter_column_block.rows());
+            }
+            else
+            {
+                LOG_DEBUG(log, "Late materialization skip read block at start_offset: {}, rows: {}", filter_column_block.startOffset(), filter_column_block.rows());
+            }
+        }
+        else
+        {
+            Block rest_column_block;
+            auto rows_left = rows - passed_count;
```

Comment:
`rows_left` means the num of rows that need to be filterd?

Reply:
`rows_left  --> filter_out_count`